File: filesync/src/main/java/de/mel/filesync/bash/BufferedIterator.java
package de.mel.filesync.bash;

import java.io.BufferedReader;
import java.io.Reader;
import java.util.NoSuchElementException;

import de.mel.auth.file.AbstractFile;

/**
 * Created by xor on 7/24/17.
 */
public abstract class BufferedIterator<T> extends BufferedReader implements AutoKlausIterator<T> {

    public abstract T convert(String line);


    public BufferedIterator(Reader in) {
        super(in);
    }

    String nextLine = null;

    @Override
    public boolean hasNext() {
        if (nextLine != null) {
            return true;
        } else {
            try {
                nextLine = readLine();
                return (nextLine != null);
            } catch (Exception e) {
                e.printStackTrace();
                return false;
            }
        }
    }

    @Override
    public T next() {
        if (nextLine != null || hasNext()) {
            String line = nextLine;
            nextLine = null;
            return convert(line);
        } else {
            throw new NoSuchElementException();
        }
    }


    public static class BufferedFileIterator extends BufferedIterator<AbstractFile> {
        public BufferedFileIterator(Reader in) {
            super(in);
        }

        @Override
        public AbstractFile convert(String line) {
            return AbstractFile.instance(line);
        }
    }

    public static class BufferedStringIterator extends BufferedIterator<String> {

        public BufferedStringIterator(Reader in) {
            super(in);
        }

        @Override
        public String convert(String line) {
            return line;
        }
    }
}

File: filesync/src/main/java/de/mel/filesync/bash/BufferedIterator.kt
package de.mel.filesync.bash

import de.mel.auth.file.AbstractFile
import de.mel.auth.file.IFile
import java.io.BufferedReader
import java.io.Reader
import java.util.*

/**
 * Created by xor on 7/24/17.
 */
abstract class BufferedIterator<T>(ins: Reader) : BufferedReader(ins), AutoKlausIterator<T> {
    abstract fun convert(line: String): T
    internal var nextLine: String? = null
    override fun hasNext(): Boolean {
        return if (nextLine != null) {
            true
        } else {
            try {
                nextLine = readLine()
                nextLine != null
            } catch (e: Exception) {
                e.printStackTrace()
                false
            }
        }
    }

    override fun next(): T {
        if (nextLine != null || hasNext()) {
            val line = nextLine!!
            nextLine = null
            return convert(line)
        } else {
            throw NoSuchElementException()
        }
    }

    class BufferedFileIterator<T : IFile>(ins: Reader) : BufferedIterator<T>(ins) {

        override fun convert(line: String): T {
            return AbstractFile.instance(line!!) as T
        }
    }

    class BufferedStringIterator(ins: Reader) : BufferedIterator<String>(ins) {
        override fun convert(line: String): String {
            return line
        }
    }
}